src/pipeline/comms/channel.rs

```diff
@@ -0,0 +1,180 @@
+use crate::pipeline::comms::{Comms, Target};
+use crate::pipeline::hashmap_thread::{HashMapCommand, HashMapHandler};
+use crate::pipeline::Result;
+use async_trait::async_trait;
+use prost::Message;
+use serde::{Deserialize, Serialize};
+use std::future::Future;
+use std::io::Cursor;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::{mpsc, oneshot};
+use tokio::{time, try_join};
+use uuid::Uuid;
+
+pub struct Channel {
+    name: &'static str,
+    next_send_chan: mpsc::Sender<Vec<u8>>,
+    prev_send_chan: mpsc::Sender<Vec<u8>>,
+    hashmap_send: mpsc::Sender<HashMapCommand>,
+    shared_id: Uuid,
+}
+impl Channel {
+    #[must_use]
+    pub fn new(
+        name: &'static str,
+        next_send_chan: mpsc::Sender<Vec<u8>>,
+        prev_send_chan: mpsc::Sender<Vec<u8>>,
+        hashmap_send: mpsc::Sender<HashMapCommand>,
+        shared_id: Uuid,
+    ) -> Channel {
+        Channel {
+            name,
+            next_send_chan,
+            prev_send_chan,
+            hashmap_send,
+            shared_id,
+        }
+    }
+
+    pub fn all_comms() -> (
+        Arc<Channel>,
+        Arc<Channel>,
+        Arc<Channel>,
+        impl Future<Output = Result<()>>,
+    ) {
+        let shared_id = Uuid::new_v4();
+        let (h1_send, h1_recv) = mpsc::channel(32);
+        let (h2_send, h2_recv) = mpsc::channel(32);
+        let (h3_send, h3_recv) = mpsc::channel(32);
+
+        let (h1_hashmap_send, h1_hashmap_recv) = mpsc::channel(32);
+        let (h2_hashmap_send, h2_hashmap_recv) = mpsc::channel(32);
+        let (h3_hashmap_send, h3_hashmap_recv) = mpsc::channel(32);
+        let h1_hashmap = HashMapHandler::new("hm1", h1_hashmap_recv);
+        let h2_hashmap = HashMapHandler::new("hm2", h2_hashmap_recv);
+        let h3_hashmap = HashMapHandler::new("hm3", h3_hashmap_recv);
+
+        let h1 = Arc::new(Channel::new(
+            "helper_1",
+            h2_send.clone(),
+            h3_send.clone(),
+            h1_hashmap_send,
+            shared_id,
+        ));
+        let h2 = Arc::new(Channel::new(
+            "helper_2",
+            h3_send.clone(),
+            h1_send.clone(),
+            h2_hashmap_send,
+            shared_id,
+        ));
+        let h3 = Arc::new(Channel::new(
+            "helper_3",
+            h1_send.clone(),
+            h2_send.clone(),
+            h3_hashmap_send,
+            shared_id,
+        ));
+        drop(h1_send);
+        drop(h2_send);
+        drop(h3_send);
```

Comment:
fair enough. I thought it would be easier to read if I kept the `clone()`s consistent, and just had the explicit `drop` at the end. but I can change. This particular code isn't terribly important, since this is for setting up MPC in a single process, which will never be used in production

Reply:
I've implemented it this way now, but it does seem a bit harder to read. you have to be aware that the `clone`s are missing in the right places